dm_control/locomotion/walkers/assets/dog_v2/add_muscles.py

```diff
@@ -0,0 +1,435 @@
+
+    Args:
+        element: The MuJoCo element for which the transformation matrix
+          is calculated.
+
+    Returns:
+        T: The transformation matrix from the root to the given element.
+  """
+  # Calculate all transformation matrices from root until this element
+  all_transformations = []
+
+  while type(element) == mjcf_module.element._ElementImpl:
+    if element.pos is not None:
+      pos = np.array(element.pos, dtype=float)
+    else:
+      pos = np.zeros(3)
+
+    if element.quat is not None:
+      rot = Rot.from_quat(element.quat).as_matrix()
+    else:
+      rot = Rot.identity().as_matrix()
+
+    all_transformations.append(
+      np.vstack(
+        (np.hstack((rot, pos.reshape([-1, 1]))), np.array([0, 0, 0, 1])))
+    )
+
+    element = element.parent
+
+  # Apply all transformations
+  T = np.eye(4)
+  for transformation in reversed(all_transformations):
+    T = np.matmul(T, transformation)
+
+  return T
+
+
+def add_muscles(model, scale_multiplier, muscle_dynamics,
+                asset_dir, lengthrange_from_joints):
+  physics = mjcf_module.Physics.from_mjcf_model(model)
+
+  mjcf = model
+
+  muscle_meshes_path = asset_dir + '/muscles/'
+  bones_path = asset_dir
+
+  bones = [f for f in listdir(bones_path)
+           if isfile(join(bones_path, f)) and f[0] != '.']
+
+  bones_kd_trees = {}
+  bones_meshes = {}
+  for bone in bones:
+    if "BONE" in bone and "simpl" not in bone and "Lingual" not in bone and "cartilage" not in bone:
+      bone_mesh = trimesh.load_mesh(bones_path + '/' + bone, process=False)
+      bones_kd_trees[bone[4:-4]] = KDTree(bone_mesh.vertices)
+      bones_meshes[bone[4:-4]] = bone_mesh
+
+  muscle = mjcf.default.default["muscle"]
+
+  flexors = muscle.add("default", dclass="flexors")
+  flexors.tendon.rgba = [0.5, 0, 0, 1]
+
+  extensors = muscle.add("default", dclass="extensors")
+  extensors.tendon.rgba = [0.5, 0, 0, 1]
+
+  length_range = mjcf.compiler.lengthrange
+  length_range.inttotal = 50
+  length_range.accel = 20
+  length_range.interval = 2
+  length_range.timestep = 0.01
+
+  mjcf.option.timestep = 0.005
+
+  muscles = extensors_front + flexors_front + \
+            extensors_back + flexors_back + torso + neck + tail
+
+  used_muscles = []
+  volumes = []
+  cross_sections = []
+  # create MTUs
+  for mtu in muscles:
+    # Load stl file of site
+    m = trimesh.load_mesh(muscle_meshes_path + mtu + '.stl')
+    pv_mesh = pv.read(muscle_meshes_path + mtu + '.stl')
+    volumes.append(pv_mesh.volume)
+    # check along which axis to slice
+    if mtu not in lateral:
+      plane_normal = [0, 0, -1]
+      start = np.argmax(m.vertices[:, 2])
+      end = np.argmin(m.vertices[:, 2])
+    else:
+      plane_normal = [1, 0, 0]
+      start = np.argmin(m.vertices[:, 0])
+      end = np.argmax(m.vertices[:, 0])
+
+    muscle_length = np.linalg.norm(m.vertices[start] - m.vertices[end])
+
+    if muscle_length < 0.1:
+      print("muscle too short", mtu)
```

Comment:
If a mucle is too short, should this give an error?

Reply:
I think it's better to notify the user that there's a muscle too short that we're not adding rather than throwing an error. There's nothing wrong with having very tiny muscle meshes, it's just that we're not using them.